Modules/_ssl.c

```diff
@@ -3289,6 +3291,200 @@ _ssl__SSLContext_load_cert_chain_impl(PySSLContext *self, PyObject *certfile,
+    ret = SSL_CTX_use_certificate(ctx, usecert);
+    if (ERR_peek_error() != 0) {
+        ret = 0;
+        goto end;
+    }
+
+    /* Certificate is set up, proceed reading chain certificates. */
+    r = SSL_CTX_clear_chain_certs(ctx);
+    if (r == 0) {
+        ret = 0;
+        goto end;
+    }
+
+    while (1) {
+        cacert = PEM_read_bio_X509(bio, NULL, pwcb, pwcb_data);
+
+        if (cacert == NULL)
+            /* Error: BIO EOF or something critical */
+            break;
+
+        r = SSL_CTX_add0_chain_cert(ctx, cacert);
+        if (!r) {
+            X509_free(cacert);
+            ret = 0;
+            goto end;
+        }
+    }
+
+    err = ERR_peek_last_error();
+    if (ERR_GET_LIB(err) == ERR_LIB_PEM
+        && ERR_GET_REASON(err) == PEM_R_NO_START_LINE)
+        /* BIO EOF, expected. */
+        ERR_clear_error();
+    else
+        /* A critical error. */
+        ret = 0;
+
+ end:
+    X509_free(usecert);
+    return ret;
+}
+
+
+/*[clinic input]
+_ssl._SSLContext._load_cert_chain_pem_from_bio
+    certbio: object(subclass_of="&PySSLMemoryBIO_Type", type="PySSLMemoryBIO *")
+    keybio: object(subclass_of="&PySSLMemoryBIO_Type", type="PySSLMemoryBIO *")
+    password: object = NULL
+
+[clinic start generated code]*/
+
+static PyObject *
+_ssl__SSLContext__load_cert_chain_pem_from_bio_impl(PySSLContext *self,
+                                                    PySSLMemoryBIO *certbio,
+                                                    PySSLMemoryBIO *keybio,
+                                                    PyObject *password)
+/*[clinic end generated code: output=1fa9f45cbdeddf1c input=4ba4aff7f210d0fd]*/
+{
+    /*
+    Note that _password_callback() starts by invoking
+
+        PySSL_END_ALLOW_THREADS_S(pw_info->thread_state);
+
+    end finishes with invoking
+
+        PySSL_BEGIN_ALLOW_THREADS_S(pw_info->thread_state);
+
+    and all OpenSSL API calls that potentially call the password callback are
+    required to be wrapped in
+
+        PySSL_BEGIN_ALLOW_THREADS_S(pw_info.thread_state);
+        [...]
+        PySSL_END_ALLOW_THREADS_S(pw_info.thread_state);
+    */
+
+    EVP_PKEY *private_key = NULL;
+    pem_password_cb *orig_passwd_cb = self->ctx->default_passwd_callback;
+
+    void *orig_passwd_userdata = self->ctx->default_passwd_callback_userdata;
+    _PySSLPasswordInfo pw_info = { NULL, NULL, NULL, 0, 0 };
+    int r;
+
+    errno = 0;
+    ERR_clear_error();
+
+    if (password && password != Py_None) {
+        if (PyCallable_Check(password)) {
+            pw_info.callable = password;
+        } else if (!_pwinfo_set(&pw_info, password,
+                                "password should be a string or callable")) {
+            goto error;
+        }
+        SSL_CTX_set_default_passwd_cb(self->ctx, _password_callback);
+        SSL_CTX_set_default_passwd_cb_userdata(self->ctx, &pw_info);
+    }
+
+
+    PySSL_BEGIN_ALLOW_THREADS_S(pw_info.thread_state);
```

Comment:
The "certificate file" is actually allowed to contain the private key. See https://docs.python.org/3/library/ssl.html#combined-key-and-certificate. One of the reasons why this function's input space is so gigantic.

In an extreme (sane) case this file contains a password-protected private key, an end-entity certificate, and N intermediate CA certificates.